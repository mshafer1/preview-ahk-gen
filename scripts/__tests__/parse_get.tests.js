const ahk_js = require('../ahk.js');

describe('_load_get', () => {
    var empty = {};
    it('returns empty for empty query', () => {
        expect(ahk_js._load_get('ahkgen.com/'))
            .toEqual(empty);
    });

    it('takes values into object', () => {
        expect(ahk_js._load_get('ahkgen.com/?length=0'))
            .toEqual({ 'length': "0" });
    });

    it('takes array values into array', () => {
        expect(ahk_js._load_get('ahkgen.com/?skey0%5B%5D=CTRL&skey0%5B%5D=ALT'))
            .toEqual({ 'skey0[]': ['CTRL', 'ALT'] })
    });

    it('takes index array and turns into array', () => {
        expect(ahk_js._load_get('ahkgen.com/?indexes%5B%5D=0'))
            .toEqual({ 'indexes[]': ['0'] })
    });

    it('takes duplicate values into array', () => {
        expect(ahk_js._load_get('ahkgen.com/?skey0=CTRL&skey0=ALT'))
            .toEqual({ 'skey0': ['CTRL', 'ALT'] })
    });

    it('takes a basic query and returns expected data', () => {
        const result = ahk_js
            ._load_get('ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=a&input0=b&option0=Replace');
        expect(result).toMatchSnapshot();
    })

    it('takes data in regardless of values', () => {
        const result = ahk_js
            ._load_get('ahkgen.com/?length=1&comment5=&func5=KEY&skeyValue5=a&input5=b&option5=Replace');
        expect(result).toMatchSnapshot();
    });

    it('takes no issue with an indexes query', () => {
        const result = ahk_js
        ._load_get(
            'ahkgen.com/?indexes%5B%5D=1&comment0=&func0=KEY&skeyValue0=a&input0=b&option0=Send'
        );
        expect(result).toMatchSnapshot();
    })
})


describe('_parse_get', () => {
    var empty = {};
    it('returns empty for empty', () => {
        expect(ahk_js._parse_get({})).toEqual(empty)
    });

    describe('takes a basic config and returns expected::', () => {
        it('returns promptly if not enough data is passed', () => {
            expect(ahk_js._parse_get({ 'length': '5' })).toHaveProperty('ERROR');
        });
        it('returns an error if length is not provided', () => {
            expect(ahk_js._parse_get({ FOO: 'bar' })).toHaveProperty('ERROR');
        });
        it('return an error for a missing index', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?indexes%5B%5D=1&comment0=&func0=KEY&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toHaveProperty('ERROR')
            expect(result).toMatchSnapshot();
        })
        it('packs values down to lowest index', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment5=&func5=KEY&skeyValue5=a&input5=b&option5=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('packs values down to lowest index', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?indexes%5B%5D=5&comment5=&func5=KEY&skeyValue5=a&input5=b&option5=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a basic send value correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        })
        it('parses a basic send value correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        })
        it('parses a basic replace correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=a&input0=b&option0=Replace'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a basic replace correctly', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skeyValue0=a&input0=b&option0=Replace'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (ctrl)', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skey0%5B%5D=CTRL&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (ctrl)', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skey0%5B%5D=CTRL&skeyValue0=a&input0=b&option0=Send'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (alt)', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skey0%5B%5D=ALT&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (alt)', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skey0%5B%5D=ALT&skeyValue0=a&input0=b&option0=Send'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (shift)', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skey0%5B%5D=SHIFT&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (shift)', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skey0%5B%5D=SHIFT&skeyValue0=a&input0=b&option0=Send'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (win)', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skey0%5B%5D=WIN&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a modifier keys correctly (win)', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skey0%5B%5D=WIN&skeyValue0=a&input0=b&option0=Send'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a group modifier keys correctly (ctrl, alt)', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skey0%5B%5D=CTRL&skey0%5B%5D=ALT&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a group modifier keys correctly (ctrl, alt)', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skey0%5B%5D=CTRL&skey0%5B%5D=ALT&skeyValue0=a&input0=b&option0=Send'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a group modifier keys correctly (ctrl, alt, shift, win)', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skey0%5B%5D=CTRL&skey0%5B%5D=SHIFT&skey0%5B%5D=ALT&skey0%5B%5D=WIN&skeyValue0=a&input0=b&option0=Send'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a group modifier keys correctly (ctrl, alt, shift, win)', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skey0%5B%5D=CTRL&skey0%5B%5D=SHIFT&skey0%5B%5D=ALT&skey0%5B%5D=WIN&skeyValue0=a&input0=b&option0=Send'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses an Activate Or Open correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=a&Window0=ahk_exe+chrome.exe&Program0=chrome.exe&option0=ActivateOrOpen'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses an Activate Or Open correctly', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skeyValue0=a&Window0=ahk_exe+chrome.exe&Program0=chrome.exe&option0=ActivateOrOpen'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a Send Unicode Char correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=b&input0=0x00A2&option0=SendUnicodeChar'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a Send Unicode Char correctly', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skeyValue0=b&input0=0x00A2&option0=SendUnicodeChar'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses an ActivateOrOpenChrome correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=b&Window0=pandora&Program0=pandora.com&option0=ActivateOrOpenChrome'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses an ActivateOrOpenChrome correctly', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skeyValue0=b&Window0=pandora&Program0=pandora.com&option0=ActivateOrOpenChrome'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses an OpenConfig correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=b&option0=OpenConfig'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses an OpenConfig correctly', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skeyValue0=b&option0=OpenConfig'
            ))
            expect(result).toMatchSnapshot();
        });
        it('parses a Custom correctly', () => {
            const result = ahk_js
                ._parse_get(ahk_js._load_get(
                    'ahkgen.com/?length=1&comment0=&func0=KEY&skeyValue0=b&Code0=send%2C+a&option0=Custom'
                ))
            expect(result).toMatchSnapshot();
        });
        it('parses a Custom correctly', () => {
            const result = ahk_js
            ._parse_get(ahk_js._load_get(
                'ahkgen.com/?indexes%5B%5D=0&comment0=&func0=KEY&skeyValue0=b&Code0=send%2C+a&option0=Custom'
            ))
            expect(result).toMatchSnapshot();
        });
    })
})
