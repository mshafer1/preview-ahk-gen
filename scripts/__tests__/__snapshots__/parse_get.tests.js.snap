// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_load_get takes a basic query and returns expected data 1`] = `
Object {
  "comment0": "",
  "func0": "KEY",
  "input0": "b",
  "length": "1",
  "option0": "Replace",
  "skeyValue0": "a",
}
`;

exports[`_load_get takes data in regardless of values 1`] = `
Object {
  "comment5": "",
  "func5": "KEY",
  "input5": "b",
  "length": "1",
  "option5": "Replace",
  "skeyValue5": "a",
}
`;

exports[`_load_get takes no issue with an indexes query 1`] = `
Object {
  "comment0": "",
  "func0": "KEY",
  "indexes[]": Array [
    "1",
  ],
  "input0": "b",
  "option0": "Send",
  "skeyValue0": "a",
}
`;

exports[`_parse_get takes a basic config and returns expected:: packs values down to lowest index 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: packs values down to lowest index 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a Custom correctly 1`] = `
Object {
  "0": Object {
    "Code": "send, a",
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "Custom",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a Custom correctly 2`] = `
Object {
  "0": Object {
    "Code": "send, a",
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "Custom",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a Send Unicode Char correctly 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "0x00A2",
    "modifiers[]": Array [],
    "option": "SendUnicodeChar",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a Send Unicode Char correctly 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "0x00A2",
    "modifiers[]": Array [],
    "option": "SendUnicodeChar",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a basic replace correctly 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [],
    "option": "Replace",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a basic replace correctly 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [],
    "option": "Replace",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a basic send value correctly 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a basic send value correctly 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a group modifier keys correctly (ctrl, alt) 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "CTRL",
      "ALT",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a group modifier keys correctly (ctrl, alt) 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "CTRL",
      "ALT",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a group modifier keys correctly (ctrl, alt, shift, win) 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "CTRL",
      "SHIFT",
      "ALT",
      "WIN",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a group modifier keys correctly (ctrl, alt, shift, win) 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "CTRL",
      "SHIFT",
      "ALT",
      "WIN",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (alt) 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "ALT",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (alt) 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "ALT",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (ctrl) 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "CTRL",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (ctrl) 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "CTRL",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (shift) 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "SHIFT",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (shift) 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "SHIFT",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (win) 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "WIN",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses a modifier keys correctly (win) 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "input": "b",
    "modifiers[]": Array [
      "WIN",
    ],
    "option": "Send",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses an Activate Or Open correctly 1`] = `
Object {
  "0": Object {
    "Program": "chrome.exe",
    "Window": "ahk_exe chrome.exe",
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "ActivateOrOpen",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses an Activate Or Open correctly 2`] = `
Object {
  "0": Object {
    "Program": "chrome.exe",
    "Window": "ahk_exe chrome.exe",
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "ActivateOrOpen",
    "skeyValue": "a",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses an ActivateOrOpenChrome correctly 1`] = `
Object {
  "0": Object {
    "Program": "pandora.com",
    "Window": "pandora",
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "ActivateOrOpenChrome",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses an ActivateOrOpenChrome correctly 2`] = `
Object {
  "0": Object {
    "Program": "pandora.com",
    "Window": "pandora",
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "ActivateOrOpenChrome",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses an OpenConfig correctly 1`] = `
Object {
  "0": Object {
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "OpenConfig",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: parses an OpenConfig correctly 2`] = `
Object {
  "0": Object {
    "func": "KEY",
    "modifiers[]": Array [],
    "option": "OpenConfig",
    "skeyValue": "b",
  },
}
`;

exports[`_parse_get takes a basic config and returns expected:: return an error for a missing index 1`] = `
Object {
  "ERROR": Object {
    "ERROR": "Missing crucial values. Must have each of func1,option1,skeyValue1",
  },
}
`;
